src/window/macos.rs

```diff
@@ -174,4 +165,53 @@ impl MacosWindowFeature {
             self.extra_titlebar_height_in_pixel
         }
     }
+
+    /// Set background color, shadow and transparency/opaqueness of the window
+    pub fn set_background(
+        &self,
+        transparency: f32,
+        show_border: bool,
+        background_color: String,
+        no_error: bool,
+    ) {
+        // Handle `g:neovide_background_color` (deprecated since 0.12.2)
+        if let Ok(color) = background_color.parse::<Color>() {
+            if !no_error {
+                error_msg!(concat!(
+                    "neovide_background_color has now been deprecated. ",
+                    "Use neovide_transparency instead if you want to get a transparent window titlebar. ",
+                    "Please check https://neovide.dev/configuration.html#background-color-deprecated-currently-macos-only for more information.",
+                ));
+            }
+            let [red, green, blue, alpha] = color.to_array();
+            unsafe {
+                let opaque = alpha >= 1.0;
+                let ns_background =
+                    NSColor::colorWithSRGBRed_green_blue_alpha(red, green, blue, alpha);
+                self.ns_window.setBackgroundColor(Some(&ns_background));
+                // If the shadow is enabled and the background color is not transparent, the window will have a grey border
+                // Workaround: Disable shadow when the background color is not transparent
+                self.ns_window.setHasShadow(opaque && show_border);
+                // Setting the window to opaque upon creation shows a permanent subtle grey border on the top edge of the window
+                self.ns_window.setOpaque(opaque && show_border);
+                self.ns_window.invalidateShadow();
+            }
+            return;
+        }
+
+        unsafe {
+            let opaque = transparency >= 1.0;
+            // Setting the background color to `NSColor::windowBackgroundColor()`
+            // makes the background opaque and draws a grey border around the window
+            let ns_background = match opaque && show_border {
+                true => NSColor::windowBackgroundColor(),
+                false => NSColor::clearColor(),
+            };
+            self.ns_window.setBackgroundColor(Some(&ns_background));
+            self.ns_window.setHasShadow(opaque);
+            // Setting the window to opaque upon creation shows a permanent subtle grey border on the top edge of the window
+            self.ns_window.setOpaque(opaque && show_border);
+            self.ns_window.invalidateShadow();
+        }
+    }
```

Comment:
honestly, this function can be improved. It's getting too long and repetitive due to the deprecated background behavior. we should probably delete the deprecated code asap. not for now.

I think we can improve the legibility creating self descriptive code.

suggestion:


```rust
/// Print a deprecation warning for `neovide_background_color`
pub fn display_deprecation_warning(&self) {
    error_msg!(concat!(
        "neovide_background_color has now been deprecated. ",
        "Use neovide_transparency instead if you want to get a transparent window titlebar. ",
        "Please check https://neovide.dev/configuration.html#background-color-deprecated-currently-macos-only for more information.",
    ));
}

#[deprecated(
    since = "0.12.2",
    note = "This function will be removed in the future."
)]
pub fn handle_legacy_background(
    &self,
    color: Color,
    show_border: bool,
    ignore_deprecation_warning: bool,
) {
    if !ignore_deprecation_warning {
        self.display_deprecation_warning();
    }
    let [red, green, blue, alpha] = color.to_array();
    unsafe {
        let opaque = alpha >= 1.0;
        let ns_background = NSColor::colorWithSRGBRed_green_blue_alpha(red, green, blue, alpha);
        self.ns_window.setBackgroundColor(Some(&ns_background));
        // If the shadow is enabled and the background color is not transparent, the window will have a grey border
        // Workaround: Disable shadow when the background color is not transparent
        self.set_window_properties(opaque, show_border);
    }
}

/// Set background color, shadow and transparency/opaqueness of the window
pub fn handle_background(&self, transparency: f32, show_border: bool) {
    unsafe {
        let opaque = transparency >= 1.0;
        // Setting the background color to `NSColor::windowBackgroundColor()`
        // makes the background opaque and draws a grey border around the window
        let ns_background = match opaque && show_border {
            true => NSColor::windowBackgroundColor(),
            false => NSColor::clearColor(),
        };
        self.ns_window.setBackgroundColor(Some(&ns_background));
        self.set_window_properties(opaque, show_border);
    }
}

/// Sets opacity and shadow properties of a window.
///
/// If `opaque` is true, the window will have a shadow and may show a grey top border.
///
/// Please note, creating the window with `opaque` will result in a subtle grey border at the top.
///
/// # Arguments
///
/// * `window` - A mutable reference to the window object
/// * `opaque` - Flag to decide if window should have shadow and potential border
/// * `show_border` - Only if `opaque == true`, this decides if a subtle grey border should be shown
pub fn set_window_properties(&self, opaque: bool, show_border: bool) {
    unsafe {
        self.ns_window.setHasShadow(opaque);
        self.ns_window.setOpaque(opaque && show_border);
        self.ns_window.invalidateShadow();
    }
}

/// Set background color, shadow and transparency/opaqueness of the window
pub fn set_background(
    &self,
    transparency: f32,
    show_border: bool,
    background_color: String,
    ignore_deprecation_warning: bool,
) {
    match background_color.parse::<Color>() {
        Ok(color) => {
            self.handle_legacy_background(color, show_border, ignore_deprecation_warning)
        }
        _ => self.handle_background(transparency, show_border),
    }
}
```

Reply:
I think it makes it harder to read because now there are three layers of function calls to set the background.
Splitting responsibility into `handle_background` and `handle_legacy_background` is a good idea but `set_window_properties` would only work for `setOpaque` and `invalidateShadow`.
The `setHasShadow` call is different for legacy and non-legacy:
```rust
// Legacy:
// If the shadow is enabled and the background color is not transparent, the window will have a grey border
// Workaround: Disable shadow when the background color is not transparent
self.ns_window.setHasShadow(opaque && show_border);
```
```rust
// Non-Legacy:
self.ns_window.setHasShadow(opaque);
```
At this point I don't think it's worth it to outsource it into a new function.
(See my new commit for the full example and let me know if you think that's good)

Also two things I don't think I quite understand:

>```rust
>/// If `opaque` is true, the window will have a shadow and may show a grey top border.
>///
>/// Please note, creating the window with `opaque` will result in a subtle grey border at the top.
>```
Aren't these two the same thing except for „creating the window with“?

>```rust
>/// # Arguments
>///
>/// * `window` - A mutable reference to the window object
>```
I don't think there is a mutable window reference argument.

Reply:
with your newest commits it looks already better.

I wasn't precise about the suggestions once it was just an idea to give you a direction.

It makes sense to refactor better later to remove th deprecated code.

it was a take it or leave it.

thank you!

Reply:
ah okay, thanks a lot for the terrific help!